rust/yarp/build.rs

```diff
@@ -0,0 +1,507 @@
+
+    writeln!(file, "#[allow(clippy::wildcard_imports)]")?;
+    writeln!(file, "use yarp_sys::*;")?;
+    writeln!(file)?;
+
+    writeln!(file, "/// A range in the source file.")?;
+    writeln!(file, "#[derive(Debug)]")?;
+    writeln!(file, "pub struct Location(*mut yp_location_t);")?;
+    writeln!(file)?;
+
+    writeln!(file, "/// A list of ranges in the source file.")?;
+    writeln!(file, "#[derive(Debug)]")?;
+    writeln!(file, "pub struct LocationList(*mut yp_location_list_t);")?;
+    writeln!(file)?;
+
+    writeln!(file, "/// An iterator over the nodes in a list.")?;
+    writeln!(file, "pub struct NodeListIter {{")?;
+    writeln!(file, "    list: *mut yp_node_list,")?;
+    writeln!(file, "    index: usize")?;
+    writeln!(file, "}}")?;
+
+    writeln!(file, "impl Iterator for NodeListIter {{")?;
+    writeln!(file, "    type Item = Node;")?;
+    writeln!(file)?;
+    writeln!(file, "    fn next(&mut self) -> Option<Self::Item> {{")?;
+    writeln!(file, "        if self.index >= unsafe {{ (*self.list).size }} {{")?;
+    writeln!(file, "            None")?;
+    writeln!(file, "        }} else {{")?;
+    writeln!(file, "            let node: *mut yp_node_t = unsafe {{ *((*self.list).nodes.add(self.index)) }};")?;
+    writeln!(file, "            self.index += 1;")?;
+    writeln!(file, "            Some(Node::new(node))")?;
+    writeln!(file, "        }}")?;
+    writeln!(file, "    }}")?;
+    writeln!(file, "}}")?;
+    writeln!(file)?;
+
+    writeln!(file, "/// A list of nodes.")?;
+    writeln!(file, "pub struct NodeList(*mut yp_node_list);")?;
+    writeln!(file)?;
+
+    writeln!(file, "impl NodeList {{")?;
+    writeln!(file, "    /// Returns an iterator over the nodes.")?;
+    writeln!(file, "    #[must_use]")?;
+    writeln!(file, "    pub const fn iter(&self) -> NodeListIter {{")?;
+    writeln!(file, "        NodeListIter {{")?;
+    writeln!(file, "            list: self.0,")?;
+    writeln!(file, "            index: 0")?;
+    writeln!(file, "        }}")?;
+    writeln!(file, "    }}")?;
+    writeln!(file, "}}")?;
+
+    writeln!(file, "impl std::fmt::Debug for NodeList {{")?;
+    writeln!(file, "    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {{")?;
+    writeln!(file, "        write!(f, \"{{:?}}\", self.iter().collect::<Vec<_>>())")?;
+    writeln!(file, "    }}")?;
+    writeln!(file, "}}")?;
+    writeln!(file)?;
+
+    writeln!(file, "/// A list of constants.")?;
+    writeln!(file, "#[derive(Debug)]")?;
+    writeln!(file, "pub struct ConstantList(*mut yp_constant_id_list_t);")?;
+    writeln!(file)?;
+
+    for node in &config.nodes {
+        writeln!(file, "const {}: u16 = yp_node_type::{} as u16;", type_name(&node.name), type_name(&node.name))?;
+    }
+    writeln!(file)?;
+
+    writeln!(file, "/// An enum representing the different kinds of nodes that can be parsed.")?;
+    writeln!(file, "pub enum Node {{")?;
```

Comment:
I think the node types in the tuple fields here should be of the related wrapper types, not of the C pointers. (screenshooting docs for context for other reviewers). i.e. I'd expect:

```rust
enum Node {
    AliasNode(AliasNode),
    // etc
}
```

...where the `AliasNode` in the tuple is the `AliasNode` struct.

![image](https://github.com/ruby/yarp/assets/142010/9bde81f9-bdc0-4286-bdc1-0a4c6124a8ec)

Reply:
Agree. Ideally, any references to `*mut yp_xyz` should only be accessible through an `as_mut_ptr` method.

Reply:
Since I'm seeing other cases... I think the general idea with this crate should be that most, if not all, interactions with pointers should be abstracted away; no one working with this crate should need to care about raw pointers.

Reply:
I think we don't want to do that because then any of the match expressions on node are going to have to copy the wrapping struct.

Reply:
if you're concerned about that, though I wouldn't be, you can mark the wrapper structs as `#[repr(transparent)]`.

As it is, this enum is able to be accessed from safe code like this:

```rust
let mut node = Node::AliasNode(
let mut uaf = Node::AliasNode(std::ptr::null_mut());
std::mem::swap(&mut node, &mut uaf);
```

and then proceed with safe operations on `node`. Does that allow for unsound usage of this crate's APIs?

The suggestion for wrapper structs means the payload of the enum is opaque. If the wrapper struct has a private constructor, then it cannot be modified by user code.